src/HTTP.jl

```diff
@@ -569,39 +568,23 @@ relationship with [`HTTP.Response`](@ref), [`HTTP.Parsers`](@ref),
 ```
 *See `docs/src/layers`[`.monopic`](http://monodraw.helftone.com).*
 """
-function stack(;redirect=true,
-                aws_authorization=false,
-                cookies=false,
-                canonicalize_headers=false,
-                retry=true,
-                status_exception=true,
-                readtimeout=0,
-                detect_content_type=false,
-                verbose=0,
-                kw...)
-
-    NoLayer = Union
-    stack =                 TopLayer{
-    (redirect             ? RedirectLayer             : NoLayer){
-                            BasicAuthLayer{
-    (detect_content_type  ? ContentTypeDetectionLayer : NoLayer){
-    (cookies === true || (cookies isa AbstractDict && !isempty(cookies)) ?
-                            CookieLayer               : NoLayer){
-    (canonicalize_headers ? CanonicalizeLayer         : NoLayer){
-                            MessageLayer{
-    (aws_authorization    ? AWS4AuthLayer             : NoLayer){
-    (retry                ? RetryLayer                : NoLayer){
-    (status_exception     ? ExceptionLayer            : NoLayer){
-                            ConnectionPoolLayer{
-    (verbose >= 3 ||
-     DEBUG_LEVEL[] >= 3   ? DebugLayer                : NoLayer){
-    (readtimeout > 0      ? TimeoutLayer              : NoLayer){
-                            StreamLayer{Union{}}
-    }}}}}}}}}}}}}
-
-    reduce(Layers.EXTRA_LAYERS; init=stack) do stack, (before, custom)
-        insert(stack, before, custom)
+function stack(; kw...)
+
+    layers = stacklayertypes(Layers.ConnectionLayer, StreamLayer(); kw...)
+    layers = ConnectionPoolLayer(layers; kw...)
+    layers = stacklayertypes(Layers.RequestLayer, layers; kw...)
+    layers = MessageLayer(layers; kw...)
+    return stacklayertypes(Layers.InitialLayer, layers; kw...)
+end
+
+function stacklayertypes(::Type{T}, layers; kw...) where {T}
+    for (k, _) in pairs(kw)
+        layer = Layers.keywordforlayer(Val(k))
```

Comment:
I am not sure if I understood the design here, but in order to insert a new layer you need to 1) overload this method, and 2) pass the matching kwarg?

Reply:
Yes, that's the idea. The only "baked in" layers then are those where the keyword args included in this `stack` function call. I've also updated the mechanism (see below) so that if the layer constructor returns `nothing` it isn't included in the stack.

Reply:
I find it slightly strange that you need to do both of these things. If you need to modify the call to HTTP.request anyway (to add the kwarg) why not pass the layer directly?

Reply:
That would be pretty inconvenient I think; it would be the difference between:
```julia
HTTP.get(...; retry=true) # includes the RetryLayer based on retry keyword "hook"
```
vs.
```julia
HTTP.get(...; layer=HTTP.RetryLayer(; retry=true))
```
There's awkwardness there because how do we "pass" the layer? We can't do `retry=HTTP.RetryLayer()` unless we loop through all keyword arguments and check if the value is a `Layer` or not. And then there's awkwardness because what if the layer author wants to use a keyword argument as an argument to their layer constructor? Then users have to be burdened with the responsibility of knowing whether to pass args/keyword args to the layer constructor or as a keyword arg to `HTTP.request`.

In my proposal, everything is keyword args and users don't have to be aware of layers at all.

Reply:
Right, but it becomes a bit of "keyword argument piracy". Right now AWS calls `HTTP.request` with their own stack I think? So in that case, even if I have loaded the aws package I won't accidentally use that layer. BrokenRecords.jl already does "layer piracy" I think?

Reply:
I think cases like AWS though shouldn't _need_ to provide an entire custom stack. They should be able to just define their own custom layer or two and then it all composes with the "base" stack in HTTP.jl. The problem with everyone providing their own stacks is there isn't any composability. There isn't a nice way to use AWS custom layer along with my own custom layer in my company's auth system.

I'm going to take some more time to study the AWS and BrokenRecord setups now that this PR is "done", because I want to make sure I understand those use-cases thoroughly and how they fit in this potential new paradigm.

Reply:
Right, the current solution isn't ideal either, just not sure the kwargs based one is better.

Reply:
Looking back over the previous layer-manipulation code vs. the proposed here and there _is_ a key important difference: the previous code led to the scenario where you were modifying a single, global, and I'll add, non-threadsafe, client request stack by inserting/removing a custom layer for a request. Ok, you weren't modifying the _whole_ stack, but it was still a single global "extra custom layers" stack. That doesn't lend itself to concurrent code needs at all!

In my proposal there _is_ no single global stack; but rather on a per request basis, and with the user-supplied keyword arguments, the stack is dynamically generated and used for the duration of that request, in isolation from any other callers.

I partly discovered this by coming back to this PR and wondering if we shouldn't just use a single global `Dict` to "register" keyword arguments instead of overloading the `Layers.keywordforlayer` with a `Val`, but then again, we could get into thread-safety issues if that keyword registry is being modified concurrently.

So while I still agree the `Layers.keywordforlayer` mechanism is somewhat ugly, it _does_ have several advantages now that make it simpler in terms of implmentation.